crates/jit/src/code_memory.rs

```diff
@@ -134,19 +134,31 @@ impl CodeMemory {
         //   both the actual unwinding tables as well as the validity of the
         //   pointers we pass in itself.
         unsafe {
+            assert!(
+                ret.text.as_ptr() as usize % region::page::size() == 0,
+                "text section is not page-aligned"
+            );
             let text_mut =
                 std::slice::from_raw_parts_mut(ret.text.as_ptr() as *mut u8, ret.text.len());
+            let mut text_section_readwrite = false;
             for (offset, r) in text.relocations() {
+                // If the text section was mapped at readonly we need to make it
+                // briefly read/write here as we apply relocations.
+                if !text_section_readwrite && self.mmap.is_readonly() {
```

Comment:
I'm actually really surprised this would work on Windows when the mapped file was opened `FILE_GENERIC_READ` as changing the protection level of the mapped pages would bypass file system access rights.